release-22.1: sql: add retries on COPY under certain conditions

otan opened the pull request:
Backport 1/1 commits from #95275.

/cc @cockroachdb/release

NOTE: this one needs a bit more care - was manually applied

---

Release note (sql change): If `copy_from_retries_enabled` is set, COPY is now able to retry certain safe circumstances - namely when `copy_from_atomic_enabled` is false, there is no transaction running COPY and the error returned is retriable. This prevents users who keep running into `TransactionProtoWithRefreshError` from having issues.

Informs #90656

Release justification: feature gated new feature critical for unblocking DMS

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95505)
<!-- Reviewable:end -->

otan committed:
sql: add retries on COPY under certain conditions

Release note (sql change): If `copy_from_retries_enabled` is set, COPY
is now able to retry certain safe circumstances - namely when
`copy_from_atomic_enabled` is false, there is no transaction running COPY
and the error returned is retriable. This prevents users who
keep running into `TransactionProtoWithRefreshError` from having issues.

cucaroach reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95505#-:-NMUYGdEDYk0TA4pv9Po:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->
